### PR TITLE
Event locations

### PR DESCRIPTION
**Pull request by rift137**
WCMS-468 Added Event addresses for off campus locations
WCMS-474 Added Event location pending
WCMS-470 Add a custom on-campus location option for events
Added campus_location model and embedded it in departments and events

Added some event specs for site categories type and set ws id

### CODE REVIEW

File: lib/buweb/campus_location.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  field :latitude, type: Float
+  field :longitude, type: Float
+  field :external_id, type: Integer # e.g. 198. From 25live
+  field :name, type: String # e.g. STHCMP ARCHERY. From 25live
+
+  validates_presence_of :formal_name, :external_id, :name
+
+  def to_s
+    formal_name
+  end
```
Comment (halloffame) [inline lib/buweb/campus_location.rb]: Can you add a method `title` method that aliases `to_s`

File: lib/buweb/event.rb [hunk 2]
```diff
@@ -28,15 +31,17 @@ class Event < Edition
...
+  embeds_one :campus_location
 
-  validates :title, presence: true
+  validates_presence_of :title, :start_date
   validates_uniqueness_of :slug, scope: :class
-  validates_presence_of :start_date
   validate :non_imported_site_presence
   validate :site_categories_type
   validate :ends_after_it_starts
+  validate :allowed_location_type
```
Comment (halloffame) [inline lib/buweb/event.rb]: You can just use this http://apidock.com/rails/ActiveModel/Validations/ClassMethods/validates_inclusion_of

File: lib/buweb/event.rb [hunk 3]
```diff
@@ -68,6 +73,10 @@ def dont_index?
     !published?
   end
 
+  def location_pending?
+    location_pending
```
Comment (halloffame) [inline lib/buweb/event.rb]: Don't you mean to say `location_type == "pending"`?

File: lib/buweb/event.rb [hunk 4]
```diff
@@ -152,9 +161,13 @@ def ends_after_it_starts
     errors.add :end_date, 'should be after the start date.' if end_date.present? && (end_date < start_date)
   end
 
+  def allowed_location_type
```
Comment (halloffame) [inline lib/buweb/event.rb]: You can remove this once you add the `validates_inclusion_of` validation

File: lib/buweb/campus_location.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  embedded_in :department
+
+  field :formal_name, type: String # e.g. Archery Range
+  field :description, type: String # e.g. By the tennis courts
+  field :latitude, type: Float
+  field :longitude, type: Float
+  field :external_id, type: Integer # e.g. 198. From 25live
+  field :name, type: String # e.g. STHCMP ARCHERY. From 25live
+
+  validates_presence_of :formal_name, :external_id, :name
```
Comment (halloffame) [inline lib/buweb/campus_location.rb]: Is there any possibility of creating campus_locations that don't come from 25live?

Comment (rift137) [inline lib/buweb/campus_location.rb]: Thats what the custom_on_campus field is going to be used for.

**Comment by halloffame** [comment]
:+1:

**Pull request merged**